packages/core/src/plugins/ConsentPlugin.ts

```diff
@@ -0,0 +1,119 @@
+import {
+  Plugin,
+  type SegmentClient,
+  type DestinationPlugin,
+  IntegrationSettings,
+  PluginType,
+  SegmentAPIIntegration,
+  SegmentEvent,
+} from '..';
+
+const SEGMENT_PREF_UPDATE_EVENT = 'Segment Consent Preference';
+
+export interface CategoryConsentStatusProvider {
+  setApplicableCategories(categories: string[]): void;
+  getConsentStatus(): Promise<Record<string, boolean>>;
+  onConsentChange(cb: (updConsent: Record<string, boolean>) => void): void;
+  shutdown?(): void;
+}
+
+/**
+ * This plugin interfaces with the consent provider and it:
+ *
+ * - stamps all events with the consent metadata.
+ * - augments all destinations with a consent filter plugin that prevents events from reaching them if
+ * they are not compliant current consent setup
+ * - listens for consent change from the provider and notifies Segment
+ */
+export class ConsentPlugin extends Plugin {
+  type = PluginType.before;
+
+  constructor(
+    private consentCategoryProvider: CategoryConsentStatusProvider,
+    private categories: string[]
+  ) {
+    super();
+  }
+
+  configure(analytics: SegmentClient): void {
+    super.configure(analytics);
+    analytics.getPlugins().forEach(this.injectConsentFilterIfApplicable);
+    analytics.onPluginLoaded(this.injectConsentFilterIfApplicable);
+    this.consentCategoryProvider.setApplicableCategories(this.categories);
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    this.consentCategoryProvider.onConsentChange(this.handleConsentChange);
+  }
+
+  async execute(event: SegmentEvent): Promise<SegmentEvent> {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if ((event as any).event === 'Segment Consent Preference') {
+      return event;
+    }
+
+    (event.context ??= {}).consent = {
+      categoryPreferences:
+        await this.consentCategoryProvider.getConsentStatus(),
```

Comment:
I think in OneTrust is not too expensive though I would still suggest using a sovran store as that gives you persistence and it would let it work offline